Fix #387, #1771: Show concept cards

BenHenning opened the pull request:
Fixes #387.
Fixes #1771.

Add support for showing concept cards via custom tags at certain points in a lesson. This introduces a new mechanism for handling custom tags. It's been integrated for both concept cards & our list tags; other tags aren't yet integrated (specifically image which is specially handled by Android's HTML parser, anyway). We will hopefully be able to leverage this to simplify #1328, as well.

This is a continuation of #422 and #731 that's been brought up-to-date with ``develop``.

This doesn't add support for concept cards in questions. #1771 is tracking that.

**Testing instructions for explorations:**
- Go to Meaning of Equal parts Exploration
- State 1-> Option D
- State 2-> 3/2 answer
- You will notice concept-card link there.

**Testing instructions for questions:**
- Go to Fractions topic
- Practice tab
- Select Mixed Numbers subtopic
- Start session
- Once you see the question "Amir says the shaded black part of the picture below represents the fraction 1/3" and select the fourth option: "No. Amir has mixed up the numerator and denominator...."
- The concept card link should show up here

BenHenning committed:
Add support for showing concept cards in feedback, and add a concept
card as one of the remediation pathways for 'the meaning of equal parts'
lesson.

BenHenning committed:
Introduce test coroutine dispatchers support in Espresso.

This piggybacks off of the solution introduced in #1276 for Robolectric.
That PR allows Robolectric tests in app module to share dependencies
with production components by creating a test application & telling
Robolectric to use it instead of OppiaApplication via a @Config
annotation.

This PR achieves the same thing by using a custom test runner that reads
the same annotation and forces Espresso & instrumentation to use the
test application instead of the default OppiaApplication. This setup
will be easier once #59 is finished since we can specify the application
in per-test manifests that both Robolectric & Espresso will respect.

Note that while this lets the same test coroutine dispatchers work in
both production & test code for Espresso, some additional work was
needed to ensure the coroutines behave correctly. In particular, the
coroutines use a fake system clock in Robolectric that requires explicit
synchronization points in the tests to allow the clock to move forward
& properly coordinate execution between background & main thread
operations. However, in Espresso, since everything is using a real clock
an idling resource is the preferred way to synchronize execution: it
allows the background coroutines to operate in real-time much like they
would in production, and then notify Espresso when completed. The test
dispatchers API is now setup to support both synchronization mechanisms
for both Robolectric & Espresso (the idling resource does nothing on
Robolectric and the force synchronization effectively does nothing on
Espresso).

The first test being demonstrated as now stable is SplashActivityTest
(as part of downstream work in #1397.

BenHenning committed:
Merge branch 'develop' into introduce-test-dispatchers-support-in-espresso

BenHenning committed:
Revert "Fixes #941: Add radar effect in Hints and solution (#1475)"

This reverts commit 41eb10bd0c04596cb18c354da7db7268121a09be.

BenHenning committed:
Merge branch 'revert-41eb10b' into introduce-test-dispatchers-support-in-espresso

BenHenning committed:
Stabilize StateFragmentTest such that it passes on both Robolectric and
Espresso.

Note that some issues were found during this: #1612 (#1611 was found a
few weeks ago, but it also affects these tests). To ensure the tests can
still be run, a @RunOn annotation was added to allow tests to target
specific test platforms. The tests that currently fail on Robolectric
due to #1611 and #1612 are disabled for that platform. The test suite as
a whole has been verified to pass in its current state on both
Robolectric and Espresso (on a Pixel XL).

The aim of this PR is to actually enable critical state fragment tests
in CI, so both StateFragmentTest and StateFragmentLocalTest are being
enabled in GitHub actions.

BenHenning committed:
Enable StateFragmentTest (Robolectric) & StateFragmentLocalTest for CI.

BenHenning committed:
Merge branch 'develop' into introduce-test-dispatchers-support-in-espresso

BenHenning committed:
Add thorough documentation for new dispatchers.

BenHenning committed:
Merge branch 'develop' into introduce-test-dispatchers-support-in-espresso

BenHenning committed:
Clean up comments & add additional documentation.

BenHenning committed:
Fix lint errors.

BenHenning committed:
Fix broken test after changes to FakeSystemClock.

BenHenning committed:
Merge branch 'introduce-test-dispatchers-support-in-espresso' into stabilize-state-fragment-test

Conflicts:
	app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt

BenHenning committed:
Merge branch 'develop' into introduce-test-dispatchers-support-in-espresso

BenHenning committed:
Merge branch 'introduce-test-dispatchers-support-in-espresso' into stabilize-state-fragment-test

BenHenning committed:
Fix linter errors.

BenHenning committed:
Merge branch 'develop' into introduce-test-dispatchers-support-in-espresso

BenHenning committed:
Merge branch 'introduce-test-dispatchers-support-in-espresso' into stabilize-state-fragment-test

BenHenning committed:
Merge branch 'develop' into stabilize-state-fragment-test

Conflicts:
	app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt

BenHenning committed:
Merge branch 'develop' into show-concept-card-links

Conflicts:
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationFragment.kt
	app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragment.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/testing/TopicTestActivity.kt
	app/src/main/java/org/oppia/app/testing/TopicTestActivityForStory.kt
	app/src/main/java/org/oppia/app/topic/TopicActivity.kt
	domain/src/main/assets/fractions_exploration1.json
	utility/src/main/java/org/oppia/util/parser/HtmlParser.kt

BenHenning committed:
Update test lesson to include references to concept cards.

BenHenning committed:
Lint fixes & use HtmlCompat instead of Html.

BenHenning committed:
Add support for the newer & finalized tag format.

BenHenning committed:
Lint fixes.

BenHenning committed:
Use a custom executor service for Glide requests that coordinates with
Oppia's test dispatchers. Note that this does not actually introduce the
service--that will happen in a new branch.

BenHenning committed:
Introduce new executor service which allows interop with Kotlin
coroutines, plus a test to verify that it fundamentally follows one
interpretation of ExecutorService's API.

BenHenning committed:
Merge branch 'develop' into introduce-coroutine-executor-service

BenHenning committed:
Fix flaky timeout tests by improving cancellation cooperation for
invokeAny() and provide longer timeouts for tests that are
CPU-sensitive.

BenHenning committed:
Add documentation & clean up unused code.

BenHenning committed:
Lint fixes.

BenHenning committed:
Significantly reorganize invokeAll() to try and make it more cooperative
for cancellation, and increase timeout times in tests to reduce
flakiness for time-sensitive tests. Some tests are remaining flaky, so
ignoring those.

Re-add maybeWithTimeoutOrNull since it actually was needed.

BenHenning committed:
Merge branch 'develop' into introduce-coroutine-executor-service

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'introduce-coroutine-executor-service' into stabilize-state-fragment-test

Conflicts:
	.github/workflows/main.yml
	app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt
	app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt

BenHenning committed:
Post-merge module fixes.

BenHenning committed:
Post-merge fixes with ratio input & add a TODO to improve speed of the
new coroutine executor service.

BenHenning committed:
Revert "Fixes part of #40 & #42: Generalisation Highfi Mobile Portrait + Landscape - Buttons (#1653)"

This reverts commit 1bb1ffa97b2c14e002fa0371991f4f25351a79cd.

BenHenning committed:
Ensure terminated tasks do not interfere with one another (timeouts
should happen individually for each task during termination). This fixes
a failure observed in StateFragmentLocalTest in #1630.

BenHenning committed:
Merge branch 'revert-1653-state-recycler-view-buttons-highfi' into introduce-coroutine-executor-service

BenHenning committed:
Merge branch 'introduce-coroutine-executor-service' into stabilize-state-fragment-test

BenHenning committed:
Ignore failing tests until #1769 is resolved.

BenHenning committed:
Fix awaitTermination & improve test. Improve stack trace for test
dispatcher timeouts.

BenHenning committed:
Merge branch 'introduce-coroutine-executor-service' into stabilize-state-fragment-test

BenHenning committed:
Fix slow & broken tests in Robolectric for StateFragmentLocalTest.

BenHenning committed:
Merge branch 'stabilize-state-fragment-test' into show-concept-card-links

BenHenning committed:
Add missing deps for StateFragmentLocalTest.

BenHenning committed:
Merge branch 'stabilize-state-fragment-test' into show-concept-card-links

BenHenning committed:
Address TODOs (including adding support for list tags which replaces the
old handler & adds nested custom tag support), and add tests.

BenHenning committed:
Lint fixes.

BenHenning commented:
I think this is ready for review. @rt4914 I need some help verifying behaviors here, in particular:
- [x] The image centering & other logic around proper image handling
- [x] The list tag management (both for newlines & the custom bullet span work); I can't actually see a difference between this being on/off

Can you help verify both of these on this branch?

anandwana001 commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -487,8 +496,6 @@ class StateFragmentPresenter @Inject constructor(
     binding.stateRecyclerView.smoothScrollToPosition(0)
   }
 
-  private fun isAudioShowing(): Boolean = viewModel.isAudioBarVisible.get()!!
```
is this mistakenly removed?

rt4914 replied:
I think this was actually not getting used anywhere.

BenHenning replied:
Correct--I removed this opportunistically.

anandwana001 commented on app/src/main/java/org/oppia/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -720,7 +730,6 @@ class StatePlayerRecyclerViewAssembler private constructor(
     private val adapterBuilder = BindableAdapter.MultiTypeBuilder.newBuilder(
       StateItemViewModel::viewType
     )
-
```
we can have this empty line as per the reformat says.

BenHenning replied:
Sure, done.

anandwana001 commented on app/src/main/java/org/oppia/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -783,10 +800,11 @@ class StatePlayerRecyclerViewAssembler private constructor(
             ).parseOppiaHtml(
               feedbackViewModel.htmlContent.toString(),
-              binding.feedbackTextView
+              binding.feedbackTextView, supportsLinks = true
```
we can have new param in the next line.

rt4914 replied:
+1

BenHenning replied:
Agreed, done. Shame that ktlint doesn't enforce this.

anandwana001 commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt:
```diff
@@ -873,6 +889,74 @@ class StateFragmentTest {
     }
   }
 
+  // TODO(BenHenning): Get these tests passing on Espresso.
```
any todo number?

BenHenning replied:
This actually needs to get done before the PR can get submitted, so I intentionally misformatted this. Keeping this comment open also as a reminder.

anandwana001 replied:
Are we good to resolve this, as you asked to keep this open as reminder?

BenHenning replied:
Ah sorry, didn't realize I was supposed to resolve it. Yep, the TODO is gone now & the tests work for me at least on Espresso.

anandwana001 commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt:
```diff
@@ -873,6 +889,74 @@ class StateFragmentTest {
+  // TODO(BenHenning): Get these tests passing on Espresso.
+
+  @Test
+  fun testStateFragment_forMisconception_showsLinkTextForConceptCard() {
```
these test case failing for me with below error

```
[Robolectric] org.oppia.app.player.state.StateFragmentTest.testStateFragment_forMisconception_clickLinkText_opensConceptCard: sdk=28; resources=BINARY
Called loadFromPath(/system/framework/framework-res.apk, true); mode=binary sdk=28

java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.oppia.testing.TestCoroutineDispatchersRobolectricImpl$RobolectricUiTaskCoordinator.idle$testing_debug(TestCoroutineDispatchersRobolectricImpl.kt:138)
	at org.oppia.testing.TestCoroutineDispatchersRobolectricImpl.flushNextTasks(TestCoroutineDispatchersRobolectricImpl.kt:91)
	at org.oppia.testing.TestCoroutineDispatchersRobolectricImpl.runCurrent(TestCoroutineDispatchersRobolectricImpl.kt:32)
	at org.oppia.app.player.state.StateFragmentTest.startPlayingExploration(StateFragmentTest.kt:972)
	at org.oppia.app.player.state.StateFragmentTest.testStateFragment_forMisconception_clickLinkText_opensConceptCard(StateFragmentTest.kt:917)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.oppia.testing.OppiaTestRule$apply$1.evaluate(OppiaTestRule.kt:18)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:546)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:252)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: GeneratedAppGlideModuleImpl is implemented incorrectly. If you've manually implemented this class, remove your implementation. The Annotation processor will generate a correct implementation.
	at com.bumptech.glide.Glide.throwIncorrectGlideModule(Glide.java:309)
	at com.bumptech.glide.Glide.getAnnotationGeneratedGlideModules(Glide.java:301)
	at com.bumptech.glide.Glide.initializeGlide(Glide.java:229)
	at com.bumptech.glide.Glide.initializeGlide(Glide.java:223)
	at com.bumptech.glide.Glide.checkAndInitializeGlide(Glide.java:184)
	at com.bumptech.glide.Glide.get(Glide.java:168)
	at com.bumptech.glide.Glide.getRetriever(Glide.java:689)
	at com.bumptech.glide.Glide.with(Glide.java:716)
	at org.oppia.util.parser.GlideImageLoader.loadBitmap(GlideImageLoader.kt:29)
	at org.oppia.util.parser.UrlImageParser.getDrawable(UrlImageParser.kt:47)
	at android.text.HtmlToSpannedConverter.startImg(Html.java:1122)
	at android.text.HtmlToSpannedConverter.handleStartTag(Html.java:835)
	at android.text.HtmlToSpannedConverter.startElement(Html.java:1209)
	at org.oppia.util.parser.CustomHtmlContentHandler.startElement(CustomHtmlContentHandler.kt:55)
	at org.ccil.cowan.tagsoup.Parser.push(Parser.java:795)
	at org.ccil.cowan.tagsoup.Parser.rectify(Parser.java:1062)
	at org.ccil.cowan.tagsoup.Parser.stagc(Parser.java:1017)
	at org.ccil.cowan.tagsoup.HTMLScanner.scan(HTMLScanner.java:624)
	at org.ccil.cowan.tagsoup.Parser.parse(Parser.java:450)
	at android.text.HtmlToSpannedConverter.convert(Html.java:746)
	at android.text.Html.fromHtml(Html.java:242)
	at androidx.core.text.HtmlCompat.fromHtml(HtmlCompat.java:162)
	at org.oppia.util.parser.CustomHtmlContentHandler$Companion.fromHtml(CustomHtmlContentHandler.kt:146)
	at org.oppia.util.parser.HtmlParser.parseOppiaHtml(HtmlParser.kt:89)
	at org.oppia.app.player.state.StatePlayerRecyclerViewAssembler$Builder$addContentSupport$2.invoke(StatePlayerRecyclerViewAssembler.kt:774)
	at org.oppia.app.player.state.StatePlayerRecyclerViewAssembler$Builder$addContentSupport$2.invoke(StatePlayerRecyclerViewAssembler.kt:722)
	at org.oppia.app.recyclerview.BindableAdapter$MultiTypeBuilder$registerViewBinder$viewHolderFactory$1$1.bind$app_debug(BindableAdapter.kt:242)
	at org.oppia.app.recyclerview.BindableAdapter.onBindViewHolder(BindableAdapter.kt:86)
	at org.oppia.app.recyclerview.BindableAdapter.onBindViewHolder(BindableAdapter.kt:30)
	at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7065)
	at androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:7107)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:6012)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6279)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6118)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6114)
	at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2303)
	at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1627)
	at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1587)
	at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:665)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4134)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:3851)
	at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4404)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1915)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at androidx.appcompat.widget.ActionBarOverlayLayout.onLayout(ActionBarOverlayLayout.java:446)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1812)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1656)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1565)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
	at com.android.internal.policy.DecorView.onLayout(DecorView.java:753)
	at android.view.View.layout(View.java:20672)
	at android.view.ViewGroup.layout(ViewGroup.java:6194)
	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2792)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2319)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1460)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7183)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:949)
	at android.view.Choreographer.doCallbacks(Choreographer.java:761)
	at android.view.Choreographer.doFrame(Choreographer.java:696)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:935)
	at android.os.Handler.handleCallback(Handler.java:873)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at org.robolectric.shadows.ShadowPausedLooper$IdlingRunnable.run(ShadowPausedLooper.java:308)
	at org.robolectric.shadows.ShadowPausedLooper.executeOnLooper(ShadowPausedLooper.java:273)
	at org.robolectric.shadows.ShadowPausedLooper.idle(ShadowPausedLooper.java:85)
	at org.oppia.testing.TestCoroutineDispatchersRobolectricImpl$RobolectricUiTaskCoordinator.idle$testing_debug(TestCoroutineDispatchersRobolectricImpl.kt:138)
	at org.oppia.testing.TestCoroutineDispatchersRobolectricImpl.flushNextTasks(TestCoroutineDispatchersRobolectricImpl.kt:91)
	at org.oppia.testing.TestCoroutineDispatchersRobolectricImpl.runCurrent(TestCoroutineDispatchersRobolectricImpl.kt:32)
	at org.oppia.app.player.state.StateFragmentTest.startPlayingExploration(StateFragmentTest.kt:972)
	at org.oppia.app.player.state.StateFragmentTest.testStateFragment_forMisconception_clickLinkText_opensConceptCard(StateFragmentTest.kt:917)
	... 15 more
Caused by: java.lang.NoSuchMethodException: com.bumptech.glide.GeneratedAppGlideModuleImpl.<init>()
	at java.lang.Class.getConstructor0(Class.java:3082)
	at java.lang.Class.getDeclaredConstructor(Class.java:2178)
	at com.bumptech.glide.Glide.getAnnotationGeneratedGlideModules(Glide.java:287)
	at com.bumptech.glide.Glide.initializeGlide(Glide.java:229)
	at com.bumptech.glide.Glide.initializeGlide(Glide.java:223)
	at com.bumptech.glide.Glide.checkAndInitializeGlide(Glide.java:184)
	at com.bumptech.glide.Glide.get(Glide.java:168)
	at com.bumptech.glide.Glide.getRetriever(Glide.java:689)
	at com.bumptech.glide.Glide.with(Glide.java:716)
	at org.oppia.util.parser.GlideImageLoader.loadBitmap(GlideImageLoader.kt:29)
	at org.oppia.util.parser.UrlImageParser.getDrawable(UrlImageParser.kt:47)
	at android.text.HtmlToSpannedConverter.$$robo$$android_text_HtmlToSpannedConverter$startImg(Html.java:1122)
	at android.text.HtmlToSpannedConverter.startImg(Html.java)
	at android.text.HtmlToSpannedConverter.$$robo$$android_text_HtmlToSpannedConverter$handleStartTag(Html.java:835)
	at android.text.HtmlToSpannedConverter.handleStartTag(Html.java)
	at android.text.HtmlToSpannedConverter.$$robo$$android_text_HtmlToSpannedConverter$startElement(Html.java:1209)
	at android.text.HtmlToSpannedConverter.startElement(Html.java)
	at org.oppia.util.parser.CustomHtmlContentHandler.startElement(CustomHtmlContentHandler.kt:55)
	at org.ccil.cowan.tagsoup.Parser.$$robo$$org_ccil_cowan_tagsoup_Parser$push(Parser.java:795)
	at org.ccil.cowan.tagsoup.Parser.push(Parser.java)
	at org.ccil.cowan.tagsoup.Parser.$$robo$$org_ccil_cowan_tagsoup_Parser$rectify(Parser.java:1062)
	at org.ccil.cowan.tagsoup.Parser.rectify(Parser.java)
	at org.ccil.cowan.tagsoup.Parser.$$robo$$org_ccil_cowan_tagsoup_Parser$stagc(Parser.java:1017)
	at org.ccil.cowan.tagsoup.Parser.stagc(Parser.java)
	at org.ccil.cowan.tagsoup.HTMLScanner.$$robo$$org_ccil_cowan_tagsoup_HTMLScanner$scan(HTMLScanner.java:624)
	at org.ccil.cowan.tagsoup.HTMLScanner.scan(HTMLScanner.java)
	at org.ccil.cowan.tagsoup.Parser.$$robo$$org_ccil_cowan_tagsoup_Parser$parse(Parser.java:450)
	at org.ccil.cowan.tagsoup.Parser.parse(Parser.java)
	at android.text.HtmlToSpannedConverter.$$robo$$android_text_HtmlToSpannedConverter$convert(Html.java:746)
	at android.text.HtmlToSpannedConverter.convert(Html.java)
	at android.text.Html.$$robo$$android_text_Html$fromHtml(Html.java:242)
	at android.text.Html.fromHtml(Html.java)
	at androidx.core.text.HtmlCompat.$$robo$$androidx_core_text_HtmlCompat$fromHtml(HtmlCompat.java:162)
	at androidx.core.text.HtmlCompat.fromHtml(HtmlCompat.java)
	at org.oppia.util.parser.CustomHtmlContentHandler$Companion.fromHtml(CustomHtmlContentHandler.kt:146)
	at org.oppia.util.parser.HtmlParser.parseOppiaHtml(HtmlParser.kt:89)
	at org.oppia.app.player.state.StatePlayerRecyclerViewAssembler$Builder$addContentSupport$2.invoke(StatePlayerRecyclerViewAssembler.kt:774)
	at org.oppia.app.player.state.StatePlayerRecyclerViewAssembler$Builder$addContentSupport$2.invoke(StatePlayerRecyclerViewAssembler.kt:722)
	at org.oppia.app.recyclerview.BindableAdapter$MultiTypeBuilder$registerViewBinder$viewHolderFactory$1$1.bind$app_debug(BindableAdapter.kt:242)
	at org.oppia.app.recyclerview.BindableAdapter.onBindViewHolder(BindableAdapter.kt:86)
	at org.oppia.app.recyclerview.BindableAdapter.onBindViewHolder(BindableAdapter.kt:30)
	at androidx.recyclerview.widget.RecyclerView$Adapter.$$robo$$androidx_recyclerview_widget_RecyclerView_Adapter$onBindViewHolder(RecyclerView.java:7065)
	at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java)
	at androidx.recyclerview.widget.RecyclerView$Adapter.$$robo$$androidx_recyclerview_widget_RecyclerView_Adapter$bindViewHolder(RecyclerView.java:7107)
	at androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java)
	at androidx.recyclerview.widget.RecyclerView$Recycler.$$robo$$androidx_recyclerview_widget_RecyclerView_Recycler$tryBindViewHolderByDeadline(RecyclerView.java:6012)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java)
	at androidx.recyclerview.widget.RecyclerView$Recycler.$$robo$$androidx_recyclerview_widget_RecyclerView_Recycler$tryGetViewHolderForPositionByDeadline(RecyclerView.java:6279)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java)
	at androidx.recyclerview.widget.RecyclerView$Recycler.$$robo$$androidx_recyclerview_widget_RecyclerView_Recycler$getViewForPosition(RecyclerView.java:6118)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java)
	at androidx.recyclerview.widget.RecyclerView$Recycler.$$robo$$androidx_recyclerview_widget_RecyclerView_Recycler$getViewForPosition(RecyclerView.java:6114)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java)
	at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.$$robo$$androidx_recyclerview_widget_LinearLayoutManager_LayoutState$next(LinearLayoutManager.java:2303)
	at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java)
	at androidx.recyclerview.widget.LinearLayoutManager.$$robo$$androidx_recyclerview_widget_LinearLayoutManager$layoutChunk(LinearLayoutManager.java:1627)
	at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java)
	at androidx.recyclerview.widget.LinearLayoutManager.$$robo$$androidx_recyclerview_widget_LinearLayoutManager$fill(LinearLayoutManager.java:1587)
	at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java)
	at androidx.recyclerview.widget.LinearLayoutManager.$$robo$$androidx_recyclerview_widget_LinearLayoutManager$onLayoutChildren(LinearLayoutManager.java:665)
	at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java)
	at androidx.recyclerview.widget.RecyclerView.$$robo$$androidx_recyclerview_widget_RecyclerView$dispatchLayoutStep2(RecyclerView.java:4134)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java)
	at androidx.recyclerview.widget.RecyclerView.$$robo$$androidx_recyclerview_widget_RecyclerView$dispatchLayout(RecyclerView.java:3851)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java)
	at androidx.recyclerview.widget.RecyclerView.$$robo$$androidx_recyclerview_widget_RecyclerView$onLayout(RecyclerView.java:4404)
	at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at androidx.constraintlayout.widget.ConstraintLayout.$$robo$$androidx_constraintlayout_widget_ConstraintLayout$onLayout(ConstraintLayout.java:1915)
	at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$onLayout(FrameLayout.java:261)
	at android.widget.FrameLayout.onLayout(FrameLayout.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$onLayout(FrameLayout.java:261)
	at android.widget.FrameLayout.onLayout(FrameLayout.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$onLayout(FrameLayout.java:261)
	at android.widget.FrameLayout.onLayout(FrameLayout.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$onLayout(FrameLayout.java:261)
	at android.widget.FrameLayout.onLayout(FrameLayout.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at androidx.appcompat.widget.ActionBarOverlayLayout.$$robo$$androidx_appcompat_widget_ActionBarOverlayLayout$onLayout(ActionBarOverlayLayout.java:446)
	at androidx.appcompat.widget.ActionBarOverlayLayout.onLayout(ActionBarOverlayLayout.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$onLayout(FrameLayout.java:261)
	at android.widget.FrameLayout.onLayout(FrameLayout.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at android.widget.LinearLayout.$$robo$$android_widget_LinearLayout$setChildFrame(LinearLayout.java:1812)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java)
	at android.widget.LinearLayout.$$robo$$android_widget_LinearLayout$layoutVertical(LinearLayout.java:1656)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java)
	at android.widget.LinearLayout.$$robo$$android_widget_LinearLayout$onLayout(LinearLayout.java:1565)
	at android.widget.LinearLayout.onLayout(LinearLayout.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$layoutChildren(FrameLayout.java:323)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java)
	at android.widget.FrameLayout.$$robo$$android_widget_FrameLayout$onLayout(FrameLayout.java:261)
	at android.widget.FrameLayout.onLayout(FrameLayout.java)
	at com.android.internal.policy.DecorView.$$robo$$com_android_internal_policy_DecorView$onLayout(DecorView.java:753)
	at com.android.internal.policy.DecorView.onLayout(DecorView.java)
	at android.view.View.$$robo$$android_view_View$layout(View.java:20672)
	at android.view.View.layout(View.java)
	at android.view.ViewGroup.$$robo$$android_view_ViewGroup$layout(ViewGroup.java:6194)
	at android.view.ViewGroup.layout(ViewGroup.java)
	at android.view.ViewRootImpl.$$robo$$android_view_ViewRootImpl$performLayout(ViewRootImpl.java:2792)
	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java)
	at android.view.ViewRootImpl.$$robo$$android_view_ViewRootImpl$performTraversals(ViewRootImpl.java:2319)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java)
	at android.view.ViewRootImpl.$$robo$$android_view_ViewRootImpl$doTraversal(ViewRootImpl.java:1460)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java)
	at android.view.ViewRootImpl$TraversalRunnable.$$robo$$android_view_ViewRootImpl_TraversalRunnable$run(ViewRootImpl.java:7183)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java)
	at android.view.Choreographer$CallbackRecord.$$robo$$android_view_Choreographer_CallbackRecord$run(Choreographer.java:949)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java)
	at android.view.Choreographer.$$robo$$android_view_Choreographer$doCallbacks(Choreographer.java:761)
	at android.view.Choreographer.doCallbacks(Choreographer.java)
	at android.view.Choreographer.$$robo$$android_view_Choreographer$doFrame(Choreographer.java:696)
	at android.view.Choreographer.doFrame(Choreographer.java)
	at android.view.Choreographer$FrameDisplayEventReceiver.$$robo$$android_view_Choreographer_FrameDisplayEventReceiver$run(Choreographer.java:935)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java)
	at android.os.Handler.$$robo$$android_os_Handler$handleCallback(Handler.java:873)
	at android.os.Handler.handleCallback(Handler.java)
	at android.os.Handler.$$robo$$android_os_Handler$dispatchMessage(Handler.java:99)
	at android.os.Handler.dispatchMessage(Handler.java)
	at org.robolectric.shadows.ShadowPausedLooper$IdlingRunnable.run(ShadowPausedLooper.java:308)
	at org.robolectric.shadows.ShadowPausedLooper.executeOnLooper(ShadowPausedLooper.java:273)
	at org.robolectric.shadows.ShadowPausedLooper.idle(ShadowPausedLooper.java:85)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.oppia.testing.TestCoroutineDispatchersRobolectricImpl$RobolectricUiTaskCoordinator.idle$testing_debug(TestCoroutineDispatchersRobolectricImpl.kt:138)
	at org.oppia.testing.TestCoroutineDispatchersRobolectricImpl.flushNextTasks(TestCoroutineDispatchersRobolectricImpl.kt:91)
	at org.oppia.testing.TestCoroutineDispatchersRobolectricImpl.runCurrent(TestCoroutineDispatchersRobolectricImpl.kt:32)
	at org.oppia.app.player.state.StateFragmentTest.startPlayingExploration(StateFragmentTest.kt:972)
	at org.oppia.app.player.state.StateFragmentTest.testStateFragment_forMisconception_clickLinkText_opensConceptCard(StateFragmentTest.kt:917)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	... 15 more
```

rt4914 replied:
This test case is passing for me.

BenHenning replied:
@anandwana001 this will happen if your Gradle dependencies are out of date (since it requires a newer Glide version). Can you try resyncing Gradle, clean + rebuilding, and running the test again to see if it still happens for you?

rt4914 commented:
**How to Test**
Go to `Meaning of Equal parts` Exploration
State 1-> Option D
State 2-> 3/2 answer
You will notice concept-card link there.

anandwana001 reviewed (changes requested):
Put a few comments

When playing what is fraction exploration after submitting 3rd question app crash with below error
```
 Process: org.oppia.app, PID: 11893
    java.lang.RuntimeException: PARAGRAPH span must end at paragraph boundary (376 follows .)
        at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:701)
        at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:674)
        at android.text.HtmlToSpannedConverter.convert(Html.java:772)
        at android.text.Html.fromHtml(Html.java:242)
        at androidx.core.text.HtmlCompat.fromHtml(HtmlCompat.java:162)
        at org.oppia.util.parser.CustomHtmlContentHandler$Companion.fromHtml(CustomHtmlContentHandler.kt:146)
        at org.oppia.util.parser.HtmlParser.parseOppiaHtml(HtmlParser.kt:89)
        at org.oppia.app.player.state.StatePlayerRecyclerViewAssembler$Builder$addContentSupport$2.invoke(StatePlayerRecyclerViewAssembler.kt:774)
        at org.oppia.app.player.state.StatePlayerRecyclerViewAssembler$Builder$addContentSupport$2.invoke(StatePlayerRecyclerViewAssembler.kt:722)
        at org.oppia.app.recyclerview.BindableAdapter$MultiTypeBuilder$registerViewBinder$viewHolderFactory$1$1.bind$app_debug(BindableAdapter.kt:242)
        at org.oppia.app.recyclerview.BindableAdapter.onBindViewHolder(BindableAdapter.kt:86)
        at org.oppia.app.recyclerview.BindableAdapter.onBindViewHolder(BindableAdapter.kt:30)
        at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7065)
```

rt4914 reviewed (approved):
LGTM, few nits and replies to @anandwana001 comments.

rt4914 reviewed (changes requested):
After checking your comment referenced to me. I am blocking this PR because of one major change in FAQ.

Regarding the comment:
1. The image centring is **correct**.
2. The li tag result has not changed. Which I think is part of of this PR, those we had a discussion that we might push this to beta but from code it looks like you did try to implement it. Am I correct or not?

The custom bullet does work correctly. 
**Steps to easily prove that CustomBullet works**
a. Go to `dimens.xml` file in **utility**. 
b. Make all values related to bullet double, as a result this should get reflected in UI.
![Screenshot_1599071480](https://user-images.githubusercontent.com/9396084/92022387-d1cdb800-ed78-11ea-9029-7154215f5c8c.png)

**No effect on Li**
![Screenshot_1599071281](https://user-images.githubusercontent.com/9396084/92022511-fd50a280-ed78-11ea-84a1-4887bbe7b66b.png)

rt4914 commented:
@BenHenning PTAL at the screenshot, how is this active even after 2 reviewers have requested changes.
<img width="1064" alt="Screenshot 2020-09-03 at 12 11 23 AM" src="https://user-images.githubusercontent.com/9396084/92023233-1c9bff80-ed7a-11ea-98fd-0df74bf99175.png">

BenHenning commented:
Ah, it's like that because it isn't based off develop. I won't merge any of the chained PRs until they're properly based on develop.

Thanks for the tip on using dimens to verify bullets--didn't think of that. One thing that confused me earlier, though: does FAQ actually use our HtmlParser today @rt4914? The bullets look different for me between the player & FAQ on develop, though the bullets aren't completely missing for me locally in FAQ.

You are correct that this means to continue support for bullet list tags, but not numbered. It should be the exact same behavior as develop per the implementation.

BenHenning committed:
Merge branch 'develop' into revert-1653-state-recycler-view-buttons-highfi

BenHenning committed:
Merge branch 'revert-1653-state-recycler-view-buttons-highfi' into introduce-coroutine-executor-service

rt4914 commented:
> Ah, it's like that because it isn't based off develop. I won't merge any of the chained PRs until they're properly based on develop.
> 
> Thanks for the tip on using dimens to verify bullets--didn't think of that. One thing that confused me earlier, though: does FAQ actually use our HtmlParser today @rt4914? The bullets look different for me between the player & FAQ on develop, though the bullets aren't completely missing for me locally in FAQ.
> 
> You are correct that this means to continue support for bullet list tags, but not numbered. It should be the exact same behavior as develop per the implementation.

Your point regarding FAQ is current. FAQ is not using HtmlParser right now. I just wrote a quick code to provide this support and the result was much better.
![Screenshot_1599074754](https://user-images.githubusercontent.com/9396084/92027574-5f60d600-ed80-11ea-9d03-5a7f5a574aee.png)
![Screenshot_1599074761](https://user-images.githubusercontent.com/9396084/92027592-64be2080-ed80-11ea-983f-3a6e26bda40b.png)


I will be creating a separate issue/PR for this fix with test.

rt4914 reviewed (approved):
LGTM, all my queries are answered and clarified. Reference: https://github.com/oppia/oppia-android/pull/1637#issuecomment-685949168

rt4914 commented:
> > Ah, it's like that because it isn't based off develop. I won't merge any of the chained PRs until they're properly based on develop.
> > Thanks for the tip on using dimens to verify bullets--didn't think of that. One thing that confused me earlier, though: does FAQ actually use our HtmlParser today @rt4914? The bullets look different for me between the player & FAQ on develop, though the bullets aren't completely missing for me locally in FAQ.
> > You are correct that this means to continue support for bullet list tags, but not numbered. It should be the exact same behavior as develop per the implementation.
> 
> Your point regarding FAQ is current. FAQ is not using HtmlParser right now. I just wrote a quick code to provide this support and the result was much better.
> ![Screenshot_1599074754](https://user-images.githubusercontent.com/9396084/92027574-5f60d600-ed80-11ea-9d03-5a7f5a574aee.png)
> ![Screenshot_1599074761](https://user-images.githubusercontent.com/9396084/92027592-64be2080-ed80-11ea-983f-3a6e26bda40b.png)
> 
> I will be creating a separate issue/PR for this fix with test.

Issue #1776 for reference.

BenHenning commented:
Thanks for checking @rt4914! In that case, I'll address the remaining comments & add the missing tests.

BenHenning committed:
Address reviewer comments.

BenHenning committed:
Merge branch 'introduce-coroutine-executor-service' into stabilize-state-fragment-test

BenHenning committed:
Merge branch 'stabilize-state-fragment-test' into show-concept-card-links

rt4914 commented:
> Put a few comments
> 
> When playing what is fraction exploration after submitting 3rd question app crash with below error
> 
> ```
>  Process: org.oppia.app, PID: 11893
>     java.lang.RuntimeException: PARAGRAPH span must end at paragraph boundary (376 follows .)
>         at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:701)
>         at android.text.SpannableStringBuilder.setSpan(SpannableStringBuilder.java:674)
>         at android.text.HtmlToSpannedConverter.convert(Html.java:772)
>         at android.text.Html.fromHtml(Html.java:242)
>         at androidx.core.text.HtmlCompat.fromHtml(HtmlCompat.java:162)
>         at org.oppia.util.parser.CustomHtmlContentHandler$Companion.fromHtml(CustomHtmlContentHandler.kt:146)
>         at org.oppia.util.parser.HtmlParser.parseOppiaHtml(HtmlParser.kt:89)
>         at org.oppia.app.player.state.StatePlayerRecyclerViewAssembler$Builder$addContentSupport$2.invoke(StatePlayerRecyclerViewAssembler.kt:774)
>         at org.oppia.app.player.state.StatePlayerRecyclerViewAssembler$Builder$addContentSupport$2.invoke(StatePlayerRecyclerViewAssembler.kt:722)
>         at org.oppia.app.recyclerview.BindableAdapter$MultiTypeBuilder$registerViewBinder$viewHolderFactory$1$1.bind$app_debug(BindableAdapter.kt:242)
>         at org.oppia.app.recyclerview.BindableAdapter.onBindViewHolder(BindableAdapter.kt:86)
>         at org.oppia.app.recyclerview.BindableAdapter.onBindViewHolder(BindableAdapter.kt:30)
>         at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7065)
> ```

Tried this and I am also getting the crash, steps that I performed.
1. Go to `What is a Fraction?`
2. Go to 3rd answer and give correct answer. 
3. Now Click continue, it will crash.

BenHenning committed:
Address review comments. Fix new concept card tests on Espresso & add
landscape versions (required configuring hints to show quickly to avoid
delaying the test, and fixing a bug in the espresso test dispatcher).
Add support for disabling concept cards of they aren't enabled for
parsing particular HTML (the default behavior is to ignore the custom
tag).

BenHenning committed:
Add support for concept cards in questions. Note that it's not clear how
to test verifying that pressing the exit button closes the concept card
since the exit button is part of the dialog's toolbar.

BenHenning commented:
I'm not hitting the crash locally. Are both of you using emulators or a real device here? @anandwana001 @rt4914

BenHenning commented:
Aha, I can hit this but only for the question where you need to pick which picture in the multiple choice list shows "3/6". I misunderstood--I interpreted '3rd question' to be the third state, not the third question being submitted.

anandwana001 commented:
> I'm not hitting the crash locally. Are both of you using emulators or a real device here? @anandwana001 @rt4914 

I had tried on emulator

BenHenning commented:
Seems like this is caused by the bullet list code. I'm able to repro it in a local test, so I should be able to find a fix.

BenHenning committed:
Fix image-breaking duplicated code in HtmlParser, fix a paragraph
parsing issue in BulletTagHandler, and add tests for
CustomHtmlContentHandler.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into introduce-coroutine-executor-service

BenHenning committed:
Merge branch 'introduce-coroutine-executor-service' into stabilize-state-fragment-test

BenHenning commented:
A lot has changed:
1. Questions are now supported (with tests)
2. Fixed the crash @anandwana001 was hitting, but there's now a potential flake in that same question. The answer image doesn't consistently load for me. @anandwana001 / @rt4914 can you verify if this is happening to you, too?
3. Added tests for the custom handler
4. A bunch of other miscellaneous fixes that were needed along the way


This should be ready for submission if the flake is localized to me or not caused by this PR.

PTAL.

rt4914 commented:
@BenHenning @anandwana001 
I haven't checked the entire code in detail. But I have done following things:
1. Checked that concept card is working in exploration and question player.
2. The image related issues are coming from develop and not from this PR. The issue is main observed we take actions in app very quickly.
3. The screenshots below are from develop.
![Screenshot_1599107062](https://user-images.githubusercontent.com/9396084/92071241-d7a0b900-edcb-11ea-86fb-ad6edd9898ac.png)
![Screenshot_1599107094](https://user-images.githubusercontent.com/9396084/92071245-da9ba980-edcb-11ea-8c6e-0ba61636aa0e.png)

BenHenning committed:
Merge branch 'develop' into stabilize-state-fragment-test

BenHenning commented:
Thanks @rt4914! I filed #1781 to track. In that case, this PR is unblocked from being able to be submitted.

BenHenning committed:
Merge branch 'stabilize-state-fragment-test' into show-concept-card-links

anandwana001 reviewed (approved):
LGTM, Thanks @BenHenning 
Below are the test results
<img width="1440" alt="Screenshot 2020-09-03 at 11 40 50" src="https://user-images.githubusercontent.com/25057618/92077603-63214680-edda-11ea-9a0b-6bd732099f33.png">

<img width="1440" alt="Screenshot 2020-09-03 at 11 43 13" src="https://user-images.githubusercontent.com/25057618/92077773-b1cee080-edda-11ea-9e96-62d5166e0a5f.png">

<img width="1440" alt="Screenshot 2020-09-03 at 11 45 08" src="https://user-images.githubusercontent.com/25057618/92077916-fa869980-edda-11ea-9b48-af39e1411e8d.png">

Espresso - QuestionPlayerActivityTest
<img width="1440" alt="Screenshot 2020-09-03 at 11 46 20" src="https://user-images.githubusercontent.com/25057618/92077981-1c801c00-eddb-11ea-8535-a7b7681bbaa6.png">

```
androidx.test.espresso.PerformException: Error performing 'single click' on view 'with id: org.oppia.app:id/selection_interaction_recyclerview'.
at androidx.test.espresso.PerformException$Builder.build(PerformException.java:86)
at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:87)
at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:59)
at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:322)
at androidx.test.espresso.ViewInteraction.desugaredPerform(ViewInteraction.java:178)
at androidx.test.espresso.ViewInteraction.perform(ViewInteraction.java:119)
at org.oppia.app.topic.questionplayer.QuestionPlayerActivityTest.clickSelection(QuestionPlayerActivityTest.kt:244)
at org.oppia.app.topic.questionplayer.QuestionPlayerActivityTest.selectMultipleChoiceOption(QuestionPlayerActivityTest.kt:232)
```

Both these test result says failure due to click issue pointing at 
```
 @Suppress("SameParameterValue")
  private fun clickSelection(optionPosition: Int, targetViewId: Int) {
    scrollToViewType(SELECTION_INTERACTION)
    onView(
      atPositionOnView(
        recyclerViewId = R.id.selection_interaction_recyclerview,
        position = optionPosition,
        targetViewId = targetViewId
      )
    ).perform(click())
    testCoroutineDispatchers.runCurrent()
  }
```

BenHenning commented:
@anandwana001 all of the tests are passing on my Pixel XL locally with Espresso. Can you provide the errors & stack traces you're seeing?

Also, can you close the old comment threads in this PR if they're resolved?

anandwana001 commented:
> @anandwana001 all of the tests are passing on my Pixel XL locally with Espresso. Can you provide the errors & stack traces you're seeing?
> 
> Also, can you close the old comment threads in this PR if they're resolved?

Resolved previous comments. 
Regarding the test, might some AS issue I have that's I they are failing. I will try running these test cases again by clean and rebuild.

anandwana001 commented:
> > @anandwana001 all of the tests are passing on my Pixel XL locally with Espresso. Can you provide the errors & stack traces you're seeing?
> > Also, can you close the old comment threads in this PR if they're resolved?
> 
> Resolved previous comments.
> Regarding the test, might some AS issue I have that's I they are failing. I will try running these test cases again by clean and rebuild.

Tried again and these two test cases failed again with the same stack trace on  Nexus 4 API 28
https://gist.github.com/anandwana001/1cd52e6f38b8e2830973f66bd4f3a7e3

Passes on Pixel XL
<img width="1440" alt="Screenshot 2020-09-03 at 12 17 18" src="https://user-images.githubusercontent.com/25057618/92080599-72ef5980-eddf-11ea-8934-b4a1bf07e734.png">

BenHenning commented:
Per discussion over chat with Akshay, it appears this is tied to testing on a small device. Akshay reported it passing on a Pixel XL. We should shore up the tests for smaller screens, but that should be done as part of a broader initiative.

BenHenning commented:
Thanks all for the reviews!
